src/magic_folder/magic_folder.py

```diff
@@ -2134,3 +2182,149 @@ def trap_conflicts(f):
             return False
         d.addErrback(trap_conflicts)
         return d.addActionFinish()
+
+
+# XXX
+# meejah: I think we discussed two queues / services: a "changed
+# file-paths" one, which produces LocalSnapshots and serializes them
+# and then a second queue that takes LocalSnapshots and uploads them
+# to the grid
+#
+# Given a list of file paths, the LocalSnapshots for these file paths
+# should be created and then persisted into the disk. A way to do that
+# would be to create a queue and push file paths into it with another
+# processing "thread" periodically emptying the queue and creating
+# LocalSnapshots and then writing them into the disk.
+# XXX: Add some state and methods to track the "status" of the upload process.
+@attr.s
+@implementer(service.IService)
+class LocalSnapshotCreator(service.Service):
+    """
+    When told about local files (that must exist in `.magic_path` or
+    below) we create a LocalSnapshot instance, serialize it and then
+    pass the instance forward to the SnapshotUploader.
+    """
+
+    magic_path = attr.ib()  # FilePath of our magic-folder base
+    db = attr.ib()  # our database
+    author = attr.ib(validator=attr.validators.instance_of(LocalAuthor))  # LocalAuthor instance
+    stash_dir = attr.ib(validator=attr.validators.instance_of(FilePath))
+    queue = attr.ib(default=attr.Factory(DeferredQueue))
+
+    def startService(self):
+        """
+        Start a periodic loop that looks for work and does it.
+        """
+        self._service_d = self._process_queue()
+
+    @inlineCallbacks
+    def _process_queue(self):
+        """
+        Wait for a single item from the queue and process it, forever.
+        """
+        while True:
+            try:
+                item = yield self.queue.get()
+                yield self._process_item(item)
+            except CancelledError:
+                break
+            except Exception:
+                write_failure(Failure())
+
+    def stopService(self):
+        """
+        Don't process queued items anymore.
+        """
+        self._service_d.cancel()
+        d = self._service_d
+        self._service_d = None
+        return d
+
+    def add_files(self, *paths):
+        """
+        Add each item in `paths` to our queue. If an item is itself a
+        directory, we add all its children. If the path does not exist
```

Comment:
Because we only process "files" in this queue. Perhaps that is the wrong design and we should process "files or directories" and ... do the recursion/walking when we "process" a directory?